IPAM ring inconsistency reconcilliation fails when range crosses zero

E.g. at https://circleci.com/gh/weaveworks/weave/12331
https://circleci.com/gh/weaveworks/weave/12345
https://circleci.com/gh/weaveworks/weave/12377

It doesn't seem to say where it's failing:

```
>>> Running /home/circleci/src/github.com/weaveworks/weave/test/191_create_bridge_2_test.sh on [test-12331-0-3.us-central1-a.weave-net test-12331-0-4.us-central1-a.weave-net]
>>> Test /home/circleci/src/github.com/weaveworks/weave/test/191_create_bridge_2_test.sh finished after 33.9 secs with error: exit status 1
Re-create bridge after restart
f66a1219fda5e35241da6aa8e2771b6f0f5876e52a8be913970e82aff7a70202
fc08a2df8e7f6a43c9b9769aec5fde74993524eb5bce3f15c8fbc27c7083b71b
ac55848b4c8b317f2179ba0a9b0364d8e9480bf04d0164df52b9a559c0347b99
d2474e5825ccd172f8ad6e2814cc87594cf725dc90ec96567f13e710b393ff5a
05234413fec5275bc5e1279bcc19d67b2b82e77f8b8da5139cd9912345edcb7b
10.32.0.2
fcf6c0c74d838a0aae215c2b70c9f241ee19c0cf6aaa324d26fb6f84a334e817
10.32.0.1
```

Looks like it happens after the `expose`.

Logs from a `weave` container after a failure like the above show it crashed; maybe this races with the `kill`?

```
INFO: 2019/08/29 16:56:12.166819 Command line options: map[docker-bridge:docker0 ipalloc-range:10.32.0.0/12 port:6783 status-addr:127.0.0.1:6782 host-root:/host nickname:test-12381-0-3 dns-listen-address:172.17.0.1:53 expect-npc:true H:[tcp://0.0.0.0:12375] datapath:datapath http-addr:127.0.0.1:6784 plugin:true proxy:true resolv-conf:/var/run/weave/etc/resolv.conf weave-bridge:weave]
INFO: 2019/08/29 16:56:12.170720 weave  unreleased
INFO: 2019/08/29 16:56:12.172623 Docker API on unix:///var/run/docker.sock: &[Arch=amd64 KernelVersion=4.13.0-1019-gcp Version=17.06.2-ce ApiVersion=1.30 GoVersion=go1.8.3 BuildTime=2017-09-05T19:59:11.118087934+00:00 MinAPIVersion=1.12 GitCommit=cec0b72 Os=linux]
INFO: 2019/08/29 16:56:12.178007 proxy listening on tcp://0.0.0.0:12375
INFO: 2019/08/29 16:56:12.274121 Bridge type is bridged_fastdp
INFO: 2019/08/29 16:56:12.274142 Communication between peers is unencrypted.
INFO: 2019/08/29 16:56:12.283837 Our name is 12:32:e3:c5:81:ce(test-12381-0-3)
INFO: 2019/08/29 16:56:12.283874 Launch detected - using supplied peer list: []
INFO: 2019/08/29 16:56:12.317998 Docker API on unix:///var/run/docker.sock: &[GoVersion=go1.8.3 BuildTime=2017-09-05T19:59:11.118087934+00:00 KernelVersion=4.13.0-1019-gcp Version=17.06.2-ce ApiVersion=1.30 MinAPIVersion=1.12 GitCommit=cec0b72 Os=linux Arch=amd64]
INFO: 2019/08/29 16:56:12.322250 Checking for pre-existing addresses on weave bridge
INFO: 2019/08/29 16:56:12.348833 [allocator 12:32:e3:c5:81:ce] No valid persisted data
INFO: 2019/08/29 16:56:12.358868 [allocator 12:32:e3:c5:81:ce] Initialising via deferred consensus
INFO: 2019/08/29 16:56:12.363968 Listening for DNS queries on 172.17.0.1:53
INFO: 2019/08/29 16:56:12.377230 Sniffing traffic on datapath (via ODP)
INFO: 2019/08/29 16:56:12.380451 Listening for HTTP control messages on 127.0.0.1:6784
INFO: 2019/08/29 16:56:12.380899 Listening for status+metrics requests on 127.0.0.1:6782
INFO: 2019/08/29 16:56:12.383485 Waiting for Weave API Server...
INFO: 2019/08/29 16:56:12.384553 Finished waiting for Weave API Server
INFO: 2019/08/29 16:56:12.384934 Listening on /run/docker/plugins/weave.sock for global scope
INFO: 2019/08/29 16:56:12.384989 Listening on /run/docker/plugins/weavemesh.sock for local scope
INFO: 2019/08/29 16:56:12.385044 Creating default "weave" network
INFO: 2019/08/29 16:56:12.402094 [ipam] RequestPool weavelocal 10.32.0.0/12  map[]
INFO: 2019/08/29 16:56:12.403617 [net] CreateNetwork c85947051287ce103fd3aa15fdf85a2739d243b9ce3b01d5c6b1475a62bd9578
INFO: 2019/08/29 16:56:13.073846 Assuming quorum size of 1
ERRO: 2019/08/29 16:56:15.811447 Interface "weave" which needs to be in UP state for Weave functioning is found to be in DOWN state
INFO: 2019/08/29 16:56:16.329137 ->[10.128.0.28:60533] connection accepted
INFO: 2019/08/29 16:56:16.329780 ->[10.128.0.28:60533|ba:2e:81:92:47:4f(test-12381-0-4)]: connection ready; using protocol version 2
INFO: 2019/08/29 16:56:16.329849 overlay_switch ->[ba:2e:81:92:47:4f(test-12381-0-4)] using sleeve
INFO: 2019/08/29 16:56:16.329878 ->[10.128.0.28:60533|ba:2e:81:92:47:4f(test-12381-0-4)]: connection added (new peer)
panic: Assertion failure

goroutine 135 [running]:
github.com/weaveworks/weave/common.Assert(0xc4204aee00)
	/go/src/github.com/weaveworks/weave/common/utils.go:8 +0x6c
github.com/weaveworks/weave/ipam.(*Allocator).assertInvariants(0xc4204aedc0)
	/go/src/github.com/weaveworks/weave/ipam/allocator.go:928 +0xd5
github.com/weaveworks/weave/ipam.(*Allocator).actorLoop(0xc4204aedc0, 0xc420470fc0, 0xc4200c8960)
	/go/src/github.com/weaveworks/weave/ipam/allocator.go:715 +0x96
created by github.com/weaveworks/weave/ipam.(*Allocator).Start
	/go/src/github.com/weaveworks/weave/ipam/allocator.go:183 +0x1e0
INFO: 2019/08/29 16:56:16.708699 Command line options: map[host-root:/host plugin:true status-addr:127.0.0.1:6782 dns-listen-address:172.17.0.1:53 docker-bridge:docker0 H:[tcp://0.0.0.0:12375] weave-bridge:weave expect-npc:true http-addr:127.0.0.1:6784 nickname:test-12381-0-3 port:6783 resolv-conf:/var/run/weave/etc/resolv.conf datapath:datapath ipalloc-range:10.32.0.0/12 proxy:true]
INFO: 2019/08/29 16:56:16.712301 weave  unreleased
INFO: 2019/08/29 16:56:16.716048 Docker API on unix:///var/run/docker.sock: &[GoVersion=go1.8.3 Os=linux Arch=amd64 BuildTime=2017-09-05T19:59:11.118087934+00:00 Version=17.06.2-ce ApiVersion=1.30 MinAPIVersion=1.12 GitCommit=cec0b72 KernelVersion=4.13.0-1019-gcp]
INFO: 2019/08/29 16:56:16.723582 proxy listening on tcp://0.0.0.0:12375
INFO: 2019/08/29 16:56:16.807601 Re-exposing 10.32.0.1/12 on bridge "weave"
INFO: 2019/08/29 16:56:16.816017 Bridge type is bridged_fastdp
INFO: 2019/08/29 16:56:16.816036 Communication between peers is unencrypted.
INFO: 2019/08/29 16:56:16.842438 Our name is 12:32:e3:c5:81:ce(test-12381-0-3)
INFO: 2019/08/29 16:56:16.842876 Restart/resume detected - using persisted peer list: []
INFO: 2019/08/29 16:56:16.854455 Docker API on unix:///var/run/docker.sock: &[Version=17.06.2-ce Os=linux GoVersion=go1.8.3 Arch=amd64 KernelVersion=4.13.0-1019-gcp BuildTime=2017-09-05T19:59:11.118087934+00:00 ApiVersion=1.30 MinAPIVersion=1.12 GitCommit=cec0b72]
INFO: 2019/08/29 16:56:16.855798 Checking for pre-existing addresses on weave bridge
INFO: 2019/08/29 16:56:16.855954 weave bridge has address 10.32.0.1/12
INFO: 2019/08/29 16:56:16.860186 [allocator 12:32:e3:c5:81:ce] Initialising with persisted data
ERRO: 2019/08/29 16:56:16.860300 [allocator] address 10.32.0.1/12 is owned by other peer ba:2e:81:92:47:4f
INFO: 2019/08/29 16:56:16.861793 Listening for DNS queries on 172.17.0.1:53
INFO: 2019/08/29 16:56:16.868207 Sniffing traffic on datapath (via ODP)
INFO: 2019/08/29 16:56:16.871609 Listening for HTTP control messages on 127.0.0.1:6784
INFO: 2019/08/29 16:56:16.872026 Listening for status+metrics requests on 127.0.0.1:6782
INFO: 2019/08/29 16:56:16.872100 Waiting for Weave API Server...
INFO: 2019/08/29 16:56:16.873004 Finished waiting for Weave API Server
INFO: 2019/08/29 16:56:16.873063 Listening on /run/docker/plugins/weave.sock for global scope
INFO: 2019/08/29 16:56:16.873099 Listening on /run/docker/plugins/weavemesh.sock for local scope
INFO: 2019/08/29 16:56:16.873107 Creating default "weave" network
ERRO: 2019/08/29 16:56:21.860686 [allocator] address 10.32.0.1/12 is owned by other peer ba:2e:81:92:47:4f
INFO: 2019/08/29 16:56:26.173182 ->[10.128.0.28:50403] connection accepted
INFO: 2019/08/29 16:56:26.173720 ->[10.128.0.28:50403|ba:2e:81:92:47:4f(test-12381-0-4)]: connection ready; using protocol version 2
INFO: 2019/08/29 16:56:26.173807 overlay_switch ->[ba:2e:81:92:47:4f(test-12381-0-4)] using sleeve
INFO: 2019/08/29 16:56:26.173837 ->[10.128.0.28:50403|ba:2e:81:92:47:4f(test-12381-0-4)]: connection added (new peer)
INFO: 2019/08/29 16:56:26.191639 ->[10.128.0.28:50403|ba:2e:81:92:47:4f(test-12381-0-4)]: connection fully established
INFO: 2019/08/29 16:56:26.192429 sleeve ->[10.128.0.28:6783|ba:2e:81:92:47:4f(test-12381-0-4)]: Effective MTU verified at 1398
INFO: 2019/08/29 16:56:45.946942 Discovered remote MAC ba:2e:81:92:47:4f at ba:2e:81:92:47:4f(test-12381-0-4)
INFO: 2019/08/29 16:56:45.947465 Discovered remote MAC b6:51:13:73:2f:2d at ba:2e:81:92:47:4f(test-12381-0-4)
```

--- a/ipam/allocator.go
+++ b/ipam/allocator.go
@@ -814,8 +814,14 @@ func (alloc *Allocator) sendRingUpdate(dest mesh.PeerName) {
 	alloc.gossip.GossipUnicast(dest, msg)
 }
 
-func (alloc *Allocator) checkRangeHasAllocations(r address.Range) bool {
-	return alloc.space.NumFreeAddressesInRange(r) != r.Size()
+func (alloc *Allocator) hasAllocations(rs []address.Range) bool {
+	for _, r := range rs {
+		alloc.debugf("hasAllocations: %v %v vs %v", r, alloc.space.NumFreeAddressesInRange(r), r.Size())
+		if alloc.space.NumFreeAddressesInRange(r) != r.Size() {
+			return true
+		}
+	}
+	return false
 }
 
 func (alloc *Allocator) update(sender mesh.PeerName, msg []byte) error {
@@ -836,7 +842,7 @@ func (alloc *Allocator) update(sender mesh.PeerName, msg []byte) error {
 	// If someone sent us a ring, merge it into ours. Note this will move us
 	// out of the awaiting-consensus state if we didn't have a ring already.
 	case data.Ring != nil:
-		updated, err := alloc.ring.Merge(*data.Ring, alloc.checkRangeHasAllocations)
+		updated, err := alloc.ring.Merge(*data.Ring, alloc.hasAllocations)
 		switch err {
 		case nil:
 			if updated {

--- a/ipam/ring/ring.go
+++ b/ipam/ring/ring.go
@@ -375,6 +375,11 @@ func (r *Ring) Empty() bool {
 	return len(r.Entries) == 0
 }
 
+// convenience function to normalise a single range that may span zero
+func (r *Ring) makeRanges(start, end address.Address) []address.Range {
+	return r.splitRangesOverZero([]address.Range{{Start: start, End: end}})
+}
+
 // Given a slice of ranges which are all in the right order except
 // possibly the last one spans zero, fix that up and return the slice
 func (r *Ring) splitRangesOverZero(ranges []address.Range) []address.Range {
@@ -425,10 +430,7 @@ type RangeInfo struct {
 
 func (r *Ring) AllRangeInfo() (result []RangeInfo) {
 	for i, entry := range r.Entries {
-		nextEntry := r.Entries.entry(i + 1)
-		ranges := []address.Range{{Start: entry.Token, End: nextEntry.Token}}
-		ranges = r.splitRangesOverZero(ranges)
-		for _, r := range ranges {
+		for _, r := range r.makeRanges(entry.Token, r.Entries.entry(i+1).Token) {
 			result = append(result, RangeInfo{entry.Peer, r, entry.Version})
 		}
 	}

--- a/ipam/ring/ring.go
+++ b/ipam/ring/ring.go
@@ -213,7 +213,7 @@ func (r *Ring) GrantRangeToHost(start, end address.Address, peer mesh.PeerName) 
 
 // Merge the given ring into this ring and indicate whether this ring
 // got updated as a result.
-func (r *Ring) Merge(gossip Ring, checkRangeHasAllocations func(r address.Range) bool) (bool, error) {
+func (r *Ring) Merge(gossip Ring, hasAllocations func(r []address.Range) bool) (bool, error) {
 	r.assertInvariants()
 	defer r.trackUpdates()()
 
@@ -238,7 +238,7 @@ func (r *Ring) Merge(gossip Ring, checkRangeHasAllocations func(r address.Range)
 		return false, ErrDifferentRange
 	}
 
-	result, updated, err := r.Entries.merge(gossip.Entries, r.Peer, r, checkRangeHasAllocations)
+	result, updated, err := r.Entries.merge(gossip.Entries, r.Peer, r, hasAllocations)
 
 	if err != nil {
 		return false, err
@@ -270,7 +270,7 @@ func (r *Ring) Merge(gossip Ring, checkRangeHasAllocations func(r address.Range)
 // entries belonging to ourPeer. Returns the merged entries and an
 // indication whether the merge resulted in any changes, i.e. the
 // result differs from the original.
-func (es entries) merge(other entries, ourPeer mesh.PeerName, r *Ring, checkRangeHasAllocations func(r address.Range) bool) (result entries, updated bool, err error) {
+func (es entries) merge(other entries, ourPeer mesh.PeerName, r *Ring, hasAllocations func(r []address.Range) bool) (result entries, updated bool, err error) {
 	var mine, theirs *entry
 	var previousOwner *mesh.PeerName
 	addToResult := func(e entry) { result = append(result, &e) }
@@ -294,7 +294,7 @@ func (es entries) merge(other entries, ourPeer mesh.PeerName, r *Ring, checkRang
 			// insert, checking that a range owned by us hasn't been split
 			if previousOwner != nil && *previousOwner == ourPeer && theirs.Peer != ourPeer {
 				// check we have no allocations in the range that got split
-				if checkRangeHasAllocations(address.Range{Start: theirs.Token, End: mine.Token}) {
+				if hasAllocations(r.makeRanges(theirs.Token, mine.Token)) {
 					err = errEntryInMyRange(theirs)
 					return
 				}
@@ -309,7 +309,7 @@ func (es entries) merge(other entries, ourPeer mesh.PeerName, r *Ring, checkRang
 				if mine.Version == theirs.Version && !mine.Equal(theirs) {
 					if mine.Peer == ourPeer {
 						// if we own the entry and has allocations
-						if checkRangeHasAllocations(address.Range{Start: mine.Token, End: es.entry(i + 1).Token}) {
+						if hasAllocations(r.makeRanges(theirs.Token, es.entry(i+1).Token)) {
 							err = errInconsistentEntry(mine, theirs)
 							return
 						}
@@ -332,7 +332,7 @@ func (es entries) merge(other entries, ourPeer mesh.PeerName, r *Ring, checkRang
 					// in the range effectively given away, or it belongs to our own peer, in
 					// which case we should set our version to the one received plus one,
 					// effectively imposing our existing entry.
-					if theirs.Peer != ourPeer && !checkRangeHasAllocations(address.Range{Start: mine.Token, End: es.entry(i + 1).Token}) {
+					if theirs.Peer != ourPeer && !hasAllocations(r.makeRanges(mine.Token, es.entry(i+1).Token)) {
 						addTheirs(*theirs)
 					} else if theirs.Peer == ourPeer {
 						mine.Version = theirs.Version + 1

--- a/ipam/ring/ring.go
+++ b/ipam/ring/ring.go
@@ -284,7 +284,7 @@ func (es entries) merge(other entries, ourPeer mesh.PeerName, r *Ring, hasAlloca
 	var i, j int
 	for i < len(es) && j < len(other) {
 		mine, theirs = es[i], other[j]
-		common.Log.Debugln(fmt.Sprintf("[ring %s]: Merge mine.Token=%s theirs.Token=%s mine.Peer=%s theirs.Peer=%s mine.Version=%s theirs.Version=%s", ourPeer, mine.Token, theirs.Token, mine.Peer, theirs.Peer, fmt.Sprint(mine.Version), fmt.Sprint(theirs.Version)))
+		common.Log.Debugf("[ring %s]: Merge mine.Token=%s theirs.Token=%s mine.Peer=%s theirs.Peer=%s mine.Version=%v theirs.Version=%v", ourPeer, mine.Token, theirs.Token, mine.Peer, theirs.Peer, mine.Version, theirs.Version)
 		switch {
 		case mine.Token < theirs.Token:
 			addToResult(*mine)
@@ -302,7 +302,6 @@ func (es entries) merge(other entries, ourPeer mesh.PeerName, r *Ring, hasAlloca
 			addTheirs(*theirs)
 			j++
 		case mine.Token == theirs.Token:
-			common.Log.Debugln(fmt.Sprintf("[ring %s]: Merge token=%s mine.Peer=%s theirs.Peer=%s mine.Version=%s theirs.Version=%s", ourPeer, mine.Token, mine.Peer, theirs.Peer, fmt.Sprint(mine.Version), fmt.Sprint(theirs.Version)))
 			// merge
 			switch {
 			case mine.Version >= theirs.Version:
@@ -333,6 +332,7 @@ func (es entries) merge(other entries, ourPeer mesh.PeerName, r *Ring, hasAlloca
 					// which case we should set our version to the one received plus one,
 					// effectively imposing our existing entry.
 					if theirs.Peer != ourPeer && !hasAllocations(r.makeRanges(mine.Token, es.entry(i+1).Token)) {
+						common.Log.Infof("ring: repair inconsistent data by accepting token %v from %v - no allocations up to %v", theirs.Token, theirs.Peer, es.entry(i+1).Token)
 						addTheirs(*theirs)
 					} else if theirs.Peer == ourPeer {
 						mine.Version = theirs.Version + 1
@@ -528,7 +528,7 @@ func (r *Ring) ReportFree(freespace map[address.Address]address.Count) (updated 
 
 		entries[i].Free = free
 		entries[i].Version++
-		common.Log.Debugln(fmt.Sprintf("[ring %s]: ReportFree token=%s peer=%s version=%s", r.Peer, entries[i].Token, entries[i].Peer, fmt.Sprint(entries[i].Version)))
+		common.Log.Debugf("[ring %s]: ReportFree token=%s peer=%s version=%v", r.Peer, entries[i].Token, entries[i].Peer, entries[i].Version)
 		updated = true
 	}
 	return
@@ -612,7 +612,7 @@ func (r *Ring) Transfer(from, to mesh.PeerName) []address.Range {
 			// bump version by large value so that transfer of range takes precedence (if there is any conflict in merging entry)
 			// over other opertaions on the entry that increments the version
 			entry.Version = entry.Version + 100
-			common.Log.Debugln(fmt.Sprintf("[ring %s]: Transfer token=%s from=%s to=%s version=%s", r.Peer, entry.Token, from, to, fmt.Sprint(entry.Version)))
+			common.Log.Debugf("[ring %s]: Transfer token=%s from=%s to=%s version=%v", r.Peer, entry.Token, from, to, entry.Version)
 			newRanges = append(newRanges, address.Range{Start: entry.Token, End: r.Entries.entry(i + 1).Token})
 		}
 	}

--- a/ipam/ring/ring_test.go
+++ b/ipam/ring/ring_test.go
@@ -38,19 +38,21 @@ func merge(r1, r2 *Ring) error {
 		}
 		return address.Count((r1.End - start) + (end - r1.Start))
 	}
-	checkEntryHasAllocations := func(r address.Range) bool {
-		size := distance(r.Start, r.End)
-		entry, found := r1.Entries.get(r.Start)
-		if !found {
-			return false
-		}
-		if entry.Free < size {
-			return true
+	check := func(rs []address.Range) bool {
+		for _, r := range rs {
+			size := distance(r.Start, r.End)
+			entry, found := r1.Entries.get(r.Start)
+			if !found {
+				continue
+			}
+			if entry.Free < size {
+				return true
+			}
 		}
 		return false
 	}
 
-	_, err := r1.Merge(*r2, checkEntryHasAllocations)
+	_, err := r1.Merge(*r2, check)
 	return err
 }
 

